test/e2e/validation_test.go

```diff
@@ -19,78 +19,94 @@ limitations under the License.
-					Name:       "infra-config1",
+				Spec: capiv1.MachineSpec{
+					ClusterName: name,
+					Bootstrap: capiv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "InfrastructureConfig",
+						Name:       "infra-config1",
+					},
 				},
-			},
-		}
+			}
 
-		err := kindClient.Create(ctx, machine)
-		Expect(err).ToNot(HaveOccurred())
+			err := kindClient.Create(ctx, machine)
+			Expect(err).ToNot(HaveOccurred())
 
-		// Confirm that object exists
-		Eventually(func() error {
-			m := &capiv1.Machine{}
-			key := client.ObjectKey{
-				Namespace: capiTestNamespace,
-				Name:      name,
-			}
-			if err := kindClient.Get(ctx, key, m); err != nil {
-				return err
-			}
-			return nil
-		}, 10*time.Second).ShouldNot(HaveOccurred())
-	})
+			// Confirm that object exists
+			Eventually(func() error {
+				m := &capiv1.Machine{}
+				key := client.ObjectKey{
+					Namespace: capiTestNamespace,
+					Name:      name,
+				}
+				if err := kindClient.Get(ctx, key, m); err != nil {
+					return err
+				}
+				return nil
+			}, 10*time.Second).ShouldNot(HaveOccurred())
+		})
 
-	It("returns error when validation fails", func() {
-		// Create invalid machine object - no bootstrap config or data.
-		name := "capi-test-machine-" + util.RandomString(6)
-		machine := &capiv1.Machine{
-			ObjectMeta: v1.ObjectMeta{
-				Name:      name,
-				Namespace: capiTestNamespace,
-			},
-			Spec: capiv1.MachineSpec{
-				ClusterName: name,
-				Bootstrap:   capiv1.Bootstrap{},
-				InfrastructureRef: corev1.ObjectReference{
-					APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
-					Kind:       "InfrastructureConfig",
-					Name:       "infra-config1",
+		It("returns error when validation fails", func() {
+			// Create invalid machine object - no bootstrap config or data.
+			name := "capi-test-machine-" + util.RandomString(6)
+			machine := &capiv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: capiTestNamespace,
 				},
-			},
-		}
+				Spec: capiv1.MachineSpec{
+					ClusterName: name,
+					Bootstrap:   capiv1.Bootstrap{},
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "InfrastructureConfig",
+						Name:       "infra-config1",
+					},
+				},
+			}
 
-		err := kindClient.Create(ctx, machine)
-		// TODO: Should we assert the error contents to ensure it is a
-		// validation error?
-		Expect(err).To(HaveOccurred())
+			err := kindClient.Create(ctx, machine)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal(
+				fmt.Sprintf(`admission webhook "validation.machine.cluster.x-k8s.io" denied the request: Machine.cluster.x-k8s.io "%s" is invalid: spec.bootstrap.data: Required value: expected spec.bootstrap.data or spec.bootstrap.configRef to be populated`, name),
+			))
```

Comment:
I tried to inspect the `*apierrors.StatusError` to see if there was a way to see if this was a "Validation" failure but there wasn't such a distinction on the error object.